client/state/signup/progress/reducer.js

```diff
@@ -3,132 +3,120 @@
 /**
  * External dependencies
  */
-import { get, find, map } from 'lodash';
+import { has, keyBy, get } from 'lodash';
 import debugFactory from 'debug';
 
 /**
  * Internal dependencies
  */
 import stepsConfig from 'signup/config/steps-pure';
-import flows from 'signup/config/flows-pure';
 import {
 	SIGNUP_COMPLETE_RESET,
 	SIGNUP_PROGRESS_COMPLETE_STEP,
 	SIGNUP_PROGRESS_INVALIDATE_STEP,
 	SIGNUP_PROGRESS_PROCESS_STEP,
-	SIGNUP_PROGRESS_REMOVE_UNNEEDED_STEPS,
 	SIGNUP_PROGRESS_RESUME_AFTER_LOGIN_SET,
 	SIGNUP_PROGRESS_SAVE_STEP,
 	SIGNUP_PROGRESS_SUBMIT_STEP,
 } from 'state/action-types';
 import { createReducer } from 'state/utils';
 import { schema } from './schema';
-import userFactory from 'lib/user';
 
 const debug = debugFactory( 'calypso:state:signup:progress:reducer' );
 
 //
-// Action handlers
+// Helper Functions
 //
-function overwriteSteps( state, { steps = [] } ) {
-	// When called without action.steps, this is basically a state reset function.
-	return Array.isArray( steps ) ? steps : [];
-}
+const addStep = ( state, step ) => {
+	debug( `Adding step ${ step.stepName }` );
 
-function completeStep( state, { step } ) {
-	return updateStep( state, { ...step, status: 'completed' } );
-}
+	return {
+		...state,
+		[ step.stepName ]: step,
+	};
+};
+
+const updateStep = ( state, newStepState ) => {
+	const { stepName, status } = newStepState;
+	const stepState = get( state, stepName );
```

Comment:
Same lodash nitpick as in other comment: `state[stepName]` does the same and is simpler. I've seen some more examples but you get the idea.